### DISCUSSION TITLE
What is the difference between `func` and `kernel`?

### DISCUSSION THREAD

**Discussion by cadop** [NVIDIA/warp#27] [Q&A] [answered]
Some questions on the difference:

The docs mention

> User functions can be called freely from within kernels inside the same module and accept arrays as inputs.

Are these functions that live on the GPU, but are not parallel? 

When would someone decide to use a func over a kernel?

Can a func be called outside of a kernel, or is it _only_ callable from a kernel?

A func must be defined in the same module means that they should be in the same file, or in the same total module that contains an `__init__.py`?

**Comment by mmacklin** [comment, marked as answer]
The main difference is just that a `wp.func()` can't be launched from Python, it can only be called from kernels or other functions.

In the latest versions you can call function objects from other Python modules directly. This is not possible with kernels which must be launched from host Python.

Cheers,
Miles

**Comment by drmikecrypto** [comment]
Building on @mmacklin's answer — a practical rule of thumb in current Warp:

- **`@wp.kernel`** — parallel entry point launched from Python with `wp.launch(..., dim=N)`. One thread per index in `dim`.
- **`@wp.func`** — device-side helper invoked *from inside* a kernel (or another func). No launch API; it runs serially per calling thread.

Use a func when you want to deduplicate math (collision response, force accumulation, etc.) without writing a separate kernel launch. Keep kernels as the top-level parallel work and funcs for reusable per-thread logic.
